Partial ANF transformation for Speedy.

nickchapman-da opened the pull request:
Partial ANF transformation for Speedy

This PR adds support to speedy for the ANF transform. The reason to transform is speed of execution. This PR is the 3rd attempt.

- The first attempt (#6440) was merged & then reverted (#6645) when we discovered it had an accidental change to the evaluation order.

- The second attempt (#6724) aimed to modify the original ANF transform to make it _safe_, by translating multi-apps to a cascade of single apps. This technique was ultimately rejected because it can cause a quadratic blowup. Attempts to revert to earlier Speedy code to support non-ANF execution floundered.

- And so we decided it would be best to make a 3rd attempt to support partial ANF by working forwards from the current no-ANF situation, rather than backwards from the full-ANF. This is that PR.

What is here:
- ANF transform code is from attempt #1 (and made _safe_ in attempt #2)
- No use of types `SExpr` and `AExpr` to distinguish before/after ANF. There is one type `SExpr` with all the original non-ANF expression forms + some new forms for ANF. And the execution engine must know how to execute both non-ANF and ANF forms.
-  It is in fact possible to switch off the ANF transformation in the speedy Compiler, and everything will still work. This is different from the first two attempts where the ANF transformation was much more invasive.
- `SBuiltin` forms are stratified into pure and effectful versions (as in attempt #1). This is to allow the special form `SELet1Builtin` to be handled to very efficiently, which turns out to be a critical aspect of the speedup of ANF.
- `SEBuiltinRecursiveDefinition` expansion bodies are recoded to use ANF forms (as in attempt #1)
- `AnfTest.scala` is from attempt #2


Here are the perf numbers for `CollectAuthority`, run 6-way interleaved vs master:

    here    58.579 ±(99.9%) 0.574 ms/op [Average]
    master  68.433 ±(99.9%) 0.747 ms/op [Average]
    here    60.459 ±(99.9%) 0.720 ms/op [Average]
    master  68.451 ±(99.9%) 0.454 ms/op [Average]
    here    59.138 ±(99.9%) 0.489 ms/op [Average]
    master  67.114 ±(99.9%) 0.452 ms/op [Average]

About x1.13 -- x1.17

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

hurryabit commented on daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/AnfTest.scala:
```diff
@@ -0,0 +1,261 @@
+    }
+  }
+
+  "unknown multi-arg fun: [\\f g. f (g 1) (g 2)]" should {
```
This test says we're transforming
```
f (g 1) g(2)
```
into
```
let x = g 1 in
let y = f x in
let z = g 2 in
y z
```
doesn't it? To me this looks exactly like the cascade of `let`s and single argument applications we want to avoid.

nickchapman-da replied:
Oh sorry. That is very embarrassing. I had the wrong version of the transform switched on.
What is worse. When I switch to the transform we intend, it crashes in just the same was as attempt #2.

nickchapman-da replied:
I need to look carefully at what is going on here.

nickchapman-da replied:
ok, so the new transform, which avoids the transform to a cascade of single-apps in now switched on
and the AnfTests are updated to show this

nickchapman-da commented:
Flipped this PR back to draft.

I now am starting to doubt the correctness of the Anf transform code. I'm not convinced the depth info is being propagated correctly. But somehow this bug didn't manifest in full-ANF

nickchapman-da commented on daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/AnfTest.scala:
```diff
@@ -0,0 +1,261 @@
+  "nested lambda: [\\f g x. g (\\y. f (f y)) x]" should {
+    "be transformed to ANF as expected" in {
+      val original =
+        lam(2, app(arg1, clo1(arg0, 1, app(free0, app(free0, arg0)))))
```
fixed the title

nickchapman-da commented on daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/AnfTest.scala:
```diff
@@ -0,0 +1,261 @@
+    }
+  }
+
+  "error (over) applied to 2 arg" should {
```
yes

nickchapman-da commented:
I fixed the bug in `transformMultiAppSafely` which was messing the shifting of stack locations
And I added a test for the correct transform in AnfTests.

nickchapman-da commented:
Updated the perf numbers

nickchapman-da commented:
rebased on master & fixed conflicts

nickchapman-da commented:
rebased on master (containing split out PR to stratify builtins)

nickchapman-da commented:
rebased on master, to get this: https://github.com/digital-asset/daml/pull/7009

nickchapman-da commented:
rebased again on master, for 2x split on PRs and improved `foldr`
